src/serde.rs

```diff
@@ -0,0 +1,153 @@
+                Kind::StringValue(str) => serializer.serialize_str(str),
+                Kind::BoolValue(bool) => serializer.serialize_bool(*bool),
+                Kind::StructValue(struc) => {
+                    let mut map = serializer.serialize_map(Some(struc.fields.len()))?;
+                    for (k, v) in struc.fields.iter() {
+                        map.serialize_entry(k, v)?;
+                    }
+                    map.end()
+                }
+                Kind::ListValue(list) => {
+                    let mut seq = serializer.serialize_seq(Some(list.values.len()))?;
+                    for v in &list.values {
+                        seq.serialize_element(v)?;
+                    }
+                    seq.end()
+                }
+            }
+        } else {
+            serializer.serialize_none()
+        }
+    }
+}
+
+impl From<serde_json::Value> for Value {
+    fn from(value: serde_json::Value) -> Self {
+        match value {
+            serde_json::Value::Null => Self {
+                kind: Some(Kind::NullValue(0)),
+            },
+            serde_json::Value::Bool(bool) => Self {
+                kind: Some(Kind::BoolValue(bool)),
+            },
+            serde_json::Value::Number(num) => {
+                if let Some(num) = num.as_i64() {
+                    Self {
+                        kind: Some(Kind::IntegerValue(num)),
+                    }
+                } else if let Some(num) = num.as_f64() {
+                    Self {
+                        kind: Some(Kind::DoubleValue(num)),
+                    }
+                } else {
+                    Self {
+                        kind: Some(Kind::DoubleValue(0.0)),
+                    }
+                }
+            }
+            serde_json::Value::String(str) => Self {
+                kind: Some(Kind::StringValue(str)),
+            },
+            serde_json::Value::Array(arr) => Self {
+                kind: Some(Kind::ListValue(ListValue {
+                    values: arr.into_iter().map(|v| v.into()).collect(),
+                })),
+            },
+            serde_json::Value::Object(obj) => Self {
+                kind: Some(Kind::StructValue(Struct {
+                    fields: obj
+                        .into_iter()
+                        .map(|(k, v)| (k, v.into()))
+                        .collect::<HashMap<_, _>>(),
+                })),
+            },
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Value {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // rely on serde_json to materialize a JSON value for conversion
+        let serde_value = serde_json::Value::deserialize(deserializer)?;
+        Ok(serde_value.into())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::client::Payload;
+
+    #[test]
+    fn json_payload_round_trip() {
```

Comment:
I've grown a bit wary of roundtrip tests, after seeing an instance where an error in the lap back masked another in the forward one. Nowadays I prefer using snapshot tests for such things (e.g. via [insta](https://docs.rs/insta).

Reply:
Let's do it in a separate PR